src/instrument.cpp

```diff
@@ -300,19 +343,153 @@ void BytecodeRewriter::rewriteCode() {
         u16 end_pc = get16();
         u16 handler_pc = get16();
         u16 catch_type = get16();
-        put16(EXTRA_BYTECODES + start_pc);
-        put16(EXTRA_BYTECODES + end_pc);
-        put16(EXTRA_BYTECODES + handler_pc);
+        put16(start_pc + relocation_table[start_pc]);
+        put16(start_pc + relocation_table[end_pc]);
+        put16(start_pc + relocation_table[handler_pc]);
         put16(catch_type);
     }
 
-    rewriteAttributes(SCOPE_REWRITE_CODE);
+    rewriteCodeAttributes(relocation_table);
+    delete[] relocation_table;
 
     // Patch attribute length
     *(u32*)(_dst + code_begin - 4) = htonl(_dst_len - code_begin);
 }
 
-void BytecodeRewriter::rewriteBytecodeTable(int data_len) {
+// Return the relocation after the last byte of code
+u32 BytecodeRewriter::rewriteCodeForLatency(const u8* code, u32 code_length, u32* relocation_table) {
+    // First scan: identify the maximum possible relocation for any position in code[]
+    // Relocation can happen due to:
+    // - Adding a new invocation
+    // - Narrow jump becoming a wide jump
+    u32 max_relocation = 0;
+    for (u32 i = 0; i < code_length;) {
+        u8 opcode = code[i];
+        if (isFunctionExit(opcode)) {
+            max_relocation += EXTRA_BYTECODES;
+        } else if (isNarrowJump(opcode)) {
+            max_relocation += 2;
+        }
+        i += OPCODE_LENGTH[opcode];
+    }
+
+    if (max_relocation > MAX_CODE_SEGMENT_BYTES - code_length) {
+        Log::warn("Instrumented code size exceeds JVM code segment size limit (%u), aborting", MAX_CODE_SEGMENT_BYTES);
+        put(code, code_length);
+        for (u32 i = 0; i < code_length; ++i) relocation_table[i] = 0;
+        return 0;
+    }
+
+    u32 code_segment_begin = _dst_len;
+
+    writeInvokeRecordSample(true);
+
+    // Method start is relocated
+    u32 current_relocation = EXTRA_BYTECODES;
+    // Low 32 bits: jump base index
+    // High 32 bits: jump offset index
+    // This supports narrow and wide jumps, as well as tableswitch and lookupswitch
+    std::vector<u64> jumps;
+    // Second scan: fill relocation_table and rewrite code.
+    // Any jump which is "close" to the narrow->wide threshold conservatively becomes
+    // a wide jump.
+    for (u32 i = 0; i < code_length;) {
+        u8 opcode = code[i];
+
+        if (isNarrowJump(opcode)) {
+            jumps.push_back((i + 1ULL) << 32 | i);
+
+            int16_t offset = (int16_t) ntohs(*(u16*)(code + i + 1));
+            if (max_relocation > std::numeric_limits<int16_t>::max() - offset) {
```

Comment:
It's the longest alternative for `0x7fff` I've ever seen.

Reply:
I think this is c++ being too verbose, in Java I would use `Short.MAX_VALUE`. But this code is going away anyways